### PR TITLE
exclusion graphing syntax

### LINKED ISSUE #1353: Cycling syntax: `T00` *after* the initial cycle point?
It would be very useful if we could specify e.g. `[[[ T00 ]]]` task after (but not including) the
initial cycle point. This is addressed one way through the `[[[ +PT6H/PT6H ]]]` approach (
see #1278), but this doesn't work if we cannot constrain the initial cycle point and have
different behaviour for particular times.

This is strongly linked to #1279, which suggests a `min(` syntax for a related problem.

Some ideas for shooting down (I dislike all of these, but the last one the least):
```
        [[[T00>]]]
```
```
        [[[T00>^]]]
```
```
        [[[icp_gt(T00)]]]
```
```
        [[[not_icp(T00)]]]
```
```
        [[[^+T00]]]
```

### PR DESCRIPTION
**Pull request by oliver-sanders**
Closes #1353 

Gaps in recurrences can now be represented with exclusions. Only one exclusion per recurrence is permitted, it must be appended to the end of a recurrence prefixed by an exclamation mark.

- Restructured and modified "Advanced Cycling" section of the docs
- Added documentation for #1038 with integer cycling

For want of a better term I have referred to gaps in recurrences as "exclusions", feel free to suggest a better alternative.

@hjoliver Please review
@benfitzpatrick Please review

### CODE REVIEW

File: doc/cug.tex [hunk 1]
```diff
@@ -2885,10 +2824,23 @@ \subsubsection{Graph Types}
...
 [interval]
 \end{lstlisting}
-For example, \lstinline=T00= is an example of \lstinline=[date-time]=, with an
+
+with example graph headings for each form being:
+
+\lstset{language=suiterc}
+\begin{lstlisting}
+[[[ R5/T00 ]]]           # Run 5 times at 00:00 every day
+[[[ R//PT1H ]]]          # Run every hour
```
Comment (benfitzpatrick) [inline doc/cug.tex]: Could you note that the `R//` is redundant, just to make it clear?

File: doc/cug.tex [hunk 2]
```diff
@@ -2885,10 +2824,23 @@ \subsubsection{Graph Types}
...
+[[[ 20000101T00Z/P1D ]]] # Run every day starting at 00:00 on 01-01-2000
```
Comment (benfitzpatrick) [inline doc/cug.tex]: Probably better to use `1st January 2000`.

File: lib/cylc/cycling/integer.py [hunk 3]
```diff
@@ -536,37 +564,77 @@ def get_point_from_expression(point_expr, context_point, is_required=False):
...
-        print ' + ' + str(point)
-        point = sequence.get_prev_point(point)
-
-    print
-    sequence1 = IntegerSequence('R/c1/P1', 1, 10)
-    sequence2 = IntegerSequence('R/c1/P1', 1, 10)
-    print sequence1 == sequence2
-    sequence2.set_offset(IntegerInterval('-P2'))
-    print sequence1 == sequence2
+class TestIntegerSequence(unittest.TestCase):
```
Comment (benfitzpatrick) [inline lib/cylc/cycling/integer.py]: I like this

File: lib/cylc/cycling/integer.py [hunk 4]
```diff
@@ -536,37 +564,77 @@ def get_point_from_expression(point_expr, context_point, is_required=False):
...
+            output.append(point)
+            point = sequence.get_prev_point(point)
+        self.assertEqual([int(out) for out in output], [10, 7, 4, 1])
+
+        # Test sequence comparison
+        sequence1 = IntegerSequence('R/1/P1', 1, 10)
+        sequence2 = IntegerSequence('R/1/P1', 1, 10)
+        self.assertEqual(sequence1, sequence2)
+        sequence2.set_offset(IntegerInterval('-P2'))
+        self.assertEqual(sequence1, sequence2)
```
Comment (benfitzpatrick) [inline lib/cylc/cycling/integer.py]: Are they supposed to be equal after an offset - I think not?

File: lib/cylc/cycling/iso8601.py [hunk 5]
```diff
@@ -344,9 +348,12 @@ def __init__(self, dep_section, context_start_point=None,
...
             assumed_time_zone=SuiteSpecifics.ASSUMED_TIME_ZONE
         )
-        self.recurrence = self.abbrev_util.parse_recurrence(recurrence_syntax)
+        self.recurrence, excl_point = self.abbrev_util.parse_recurrence(
+            recurrence_syntax)
+        self.exclusion = ISO8601Point.from_nonstandard_string(
+                str(excl_point)) if excl_point else None
         self.step = ISO8601Interval(str(self.recurrence.duration))
-        self.value = str(self.recurrence)
+        self.value = str(self.recurrence) + '!' + str(self.exclusion)
```
Comment (benfitzpatrick) [inline lib/cylc/cycling/iso8601.py]: You only want the `!` when there's really an exclusion

File: lib/cylc/cycling/iso8601.py [hunk 6]
```diff
@@ -366,11 +373,13 @@ def set_offset(self, offset):
         self._cached_next_point_values = {}
         self._cached_valid_point_booleans = {}
         self._cached_recent_valid_points = []
-        self.value = str(self.recurrence)
+        self.value = str(self.recurrence) + '!' + str(self.exclusion)
 
     def is_on_sequence(self, point):
         """Return True if point is on-sequence."""
         # Iterate starting at recent valid points, for speed.
+        if point == self.exclusion:
```
Comment (benfitzpatrick) [inline lib/cylc/cycling/iso8601.py]: You weren't tempted to go for the same `if exclusion and point == self.exclusion` logic in integer.py?

File: lib/cylc/cycling/iso8601.py [hunk 7]
```diff
@@ -408,16 +417,20 @@ def get_prev_point(self, point):
...
+            if res == self.exclusion:
+                return self.get_prev_point(res)
         return res
 
     def get_nearest_prev_point(self, point):
         """Return the largest point < some arbitrary point."""
         if self.is_on_sequence(point):
             return self.get_prev_point(point)
         p_iso_point = point_parse(point.value)
+        p_iso_excl = point_parse(self.exclusion.value)
```
Comment (benfitzpatrick) [inline lib/cylc/cycling/iso8601.py]: Will this break when self.exclusion is None?

File: lib/cylc/cycling/iso8601.py [hunk 8]
```diff
@@ -791,33 +825,97 @@ def _get_old_strptime_format(point_string):
...
+
+        output = []
+        point = sequence.get_start_point()
+        count = 0
+        while point and count < 4:
+            output.append(point)
+            point = sequence.get_next_point(point)
+            count += 1
+        output = [str(out) for out in output]
+        self.assertEqual(output, ['20000101T0100+01', '20000101T0200+01',
```
Comment (benfitzpatrick) [inline lib/cylc/cycling/iso8601.py]: To get these in `Z` (and hence passing elsewhere) you need to set up the time zone in the configuration singleton in this file.

File: lib/cylc/cycling/integer.py (hunk 4 above)
Comment (oliver-sanders) [inline lib/cylc/cycling/integer.py]: They should be equal as the offset is a multiple of the interval. I've added another test to cover the case where the offset is not a multiple.

File: lib/cylc/cycling/iso8601.py (hunk 8 above)
Comment (oliver-sanders) [inline lib/cylc/cycling/iso8601.py]: Done.

**Comment by hjoliver** [comment]
I'm deferring to Ben on this one!  I'll take a look once he's satisfied.

File: lib/cylc/cycling/iso8601.py [hunk 9]
```diff
@@ -344,9 +348,14 @@ def __init__(self, dep_section, context_start_point=None,
             custom_point_parse_function=self.custom_point_parse_function,
             assumed_time_zone=SuiteSpecifics.ASSUMED_TIME_ZONE
         )
-        self.recurrence = self.abbrev_util.parse_recurrence(recurrence_syntax)
+        self.recurrence, excl_point = self.abbrev_util.parse_recurrence(
+            recurrence_syntax)
+        self.exclusion = ISO8601Point.from_nonstandard_string(
+                str(excl_point)) if excl_point else None
```
Comment (benfitzpatrick) [inline lib/cylc/cycling/iso8601.py]: The pep8 test is upset here

**Comment by benfitzpatrick** [comment]
Tests pass, over to @hjoliver

**Comment by benfitzpatrick** [comment]
(Except pep8)

**Comment by oliver-sanders** [comment]
Now featuring pep8 compliance.

**Comment by oliver-sanders** [comment]
Resolved conflicts.

**Comment by hjoliver** [comment]
Looking good, review mostly done...

File: doc/cug.tex [hunk 10]
```diff
@@ -2780,87 +2780,26 @@ \subsubsection{Graph Types}
 \label{fig-test2}
 \end{figure}
 
-\subparagraph{Advanced Cycling}
-\label{AdvancedCycling}
-
-The hour-based section heading examples shown above (\lstinline=T00=,
-\lstinline=T06=, etc) are just one small part of cylc's cycling syntax.
+\subsubsection{Graph Headers}
```
Comment (hjoliver) [inline doc/cug.tex]: "Graph Section Headings"?

File: doc/cug.tex [hunk 11]
```diff
@@ -2780,87 +2780,26 @@ \subsubsection{Graph Types}
...
 
-\lstinline=T06= stands for "Run every day starting at 06:00 after the
-initial cycle point". Cylc allows you to start (or end) at any particular
-time, repeat at whatever frequency you like, and even optionally limit the
-number of repetitions.
+In the above examples graph headings are used to define recurring date-times
+when tasks are to run. For example in the following scenario,
```
Comment (hjoliver) [inline doc/cug.tex]: I would say something like, "Graph section headings define the recurrence expressions (date-time or integer sequences, depending on the suite cycling mode) for their dependency graphs.".

File: doc/cug.tex [hunk 12]
```diff
@@ -2780,87 +2780,26 @@ \subsubsection{Graph Types}
...
-        # Run every day starting at 20140201T06
-        [[[ 20140201T06/P1D ]]]
-            graph = "absolute_daily_ham"
-
-        # Run once at the first instance of either T00, T06, T12 or T18
-        # starting at the initial cycle point
-        [[[ R1/min(T00,T06,T12,T18) ]]]
-            graph = "cold_cycle_dependent_foo => foo"
-
+        [[[ T06 ]]]  # A graph heading
```
Comment (hjoliver) [inline doc/cug.tex]: "A graph section heading"

File: doc/cug.tex [hunk 13]
```diff
@@ -2780,87 +2780,26 @@ \subsubsection{Graph Types}
...
+            graph = foo => bar
 \end{lstlisting}
 \lstset{language=transcript}
 
-Here is a quick summary of the syntax rules:
+\lstinline=T06= stands for "Run every day starting at 06:00 after the
```
Comment (hjoliver) [inline doc/cug.tex]: "means" rather than "stands for"?

File: doc/cug.tex [hunk 14]
```diff
@@ -2977,15 +2914,42 @@ \subsubsection{Graph Types}
...
 \end{lstlisting}
 \lstset{language=transcript}
 
+\paragraph{Excluding Dates}
+\label{excluding-dates}
+\lstset{language=suiterc}
 
+It can be desirable to exclude a date from a recurrence, this can be achieved
+by appending the date to the recurrence prefixed with an exclamation mark.
+For example \lstinline=[[[ PT1D!20000101 ]]]= means run daily except on the
```
Comment (hjoliver) [inline doc/cug.tex]: Could be more concise, e.g.: "You can use an exclamation mark to exclude a particular date from a recurrence.  For example..."

File: doc/cug.tex [hunk 15]
```diff
@@ -2977,15 +2914,42 @@ \subsubsection{Graph Types}
...
+first of January 2000.
 
-\subparagraph{How Multiple Graph Strings Combine}
+This syntax can only be used to exclude one datetime from a recurrence, note
```
Comment (hjoliver) [inline doc/cug.tex]: Break sentence: "...recurrence.  Note..."

File: doc/cug.tex [hunk 16]
```diff
@@ -2977,15 +2914,42 @@ \subsubsection{Graph Types}
...
+that the \lstinline=^= and \lstinline=$= symbols (shorthand for the initial
+and final cycle points) are both datetimes so \lstinline=[[[ T12!$-PT1D ]]]=
+is valid.
+
+If using a run limit in combination with an exclusion, the heading might not
+run the number of times specified in the limit. For example in the following
+suite \lstinline=foo= will only run once as what would be its second run has
```
Comment (hjoliver) [inline doc/cug.tex]: "foo will only run once as its second run has ..."

File: doc/cug.tex [hunk 17]
```diff
@@ -3281,6 +3265,37 @@ \subsubsection{Graph Types}
...
+[[[ R1/$ ]]]     # Run once at the final cycle point
+[[[ R3/^/P2 ]]]  # Run three times with step two starting at the
+                 # initial cycle point
+\end{lstlisting}
+\lstset{language=transcript}
+
+Likewise the syntax introduced in~\ref{excluding-dates} for excluding
+a particular point from a recurrence also works for integer cycling
+meaning you can write:
+
```
Comment (hjoliver) [inline doc/cug.tex]: IMO, we don't need "meaning you can write" on the end of the sentence.

File: doc/cug.tex [hunk 18]
```diff
@@ -3281,6 +3265,37 @@ \subsubsection{Graph Types}
...
+\begin{lstlisting}
+[[[ R/P4!8 ]]]       # Run with step four, to the final cycle point
+                     # but not at point 8
+[[[ R3/3/P2!5 ]]]    # Run with step two from point 3 but not at
+                     # point 5
+[[[ R/+P1/P6!14 ]]]  # Run with step six from one step after the
+                     # initial cycle point but not at point 14
+\end{lstlisting}
+\lstset{language=transcript}
+
```
Comment (hjoliver) [inline doc/cug.tex]: Inconsistent use of english vs digit numbers (just use digits?) - i.e. "step 2" rather than "step two"?

**Comment by hjoliver** [comment]
Your iso8601 unit tests fail here (NZ) for timezone reasons.  (Generally nice though, I've been thinking for ages we should really do simple unit tests wherever possible as well as the full test battery...)

File: tests/cyclers/exclusions/reference.log [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+2016-07-04T09:37:28Z INFO - [qux.20000102T1200Z] -(current:running)> succeeded at 2016-07-04T09:37:27Z
+2016-07-04T09:37:28Z INFO - [nip.20000102T0600Z] -('job-logs-register', 1) will run after P0Y (after 2016-07-04T09:37:28Z)
+2016-07-04T09:37:28Z INFO - [qux.20000102T1200Z] -('job-logs-register', 1) will run after P0Y (after 2016-07-04T09:37:28Z)
+2016-07-04T09:37:28Z INFO - [client-command] get_cylc_version osanders@eld668.desktop.frd.metoffice.com:cylc-gui 3385400d-7efc-449b-a6ac-a8698a23101d
+2016-07-04T09:37:29Z INFO - [nip.20000102T1200Z] -submit_method_id=18613
+2016-07-04T09:37:29Z INFO - [nip.20000102T1200Z] -submission succeeded
+2016-07-04T09:37:29Z INFO - [nip.20000102T1200Z] -(current:submitted)> started at 2016-07-04T09:37:28Z
+2016-07-04T09:37:29Z INFO - [nip.20000102T1200Z] -(current:running)> succeeded at 2016-07-04T09:37:29Z
+2016-07-04T09:37:29Z INFO - [nip.20000102T1200Z] -('job-logs-register', 1) will run after P0Y (after 2016-07-04T09:37:29Z)
+2016-07-04T09:37:30Z INFO - Suite shutting down at 2016-07-04T09:37:30Z
```
Comment (hjoliver) [inline tests/cyclers/exclusions/reference.log]: Note we're now removing the cruft from reference logs: https://github.com/cylc/cylc/pull/1892#issuecomment-226602511  (not automated yet, unfortunately).

File: tests/cyclers/exclusions/suite.rc [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+
+        # Don't run at the initial cycle point
+        [[[ T00!^ ]]]
+            graph = bar
+
+        # Don't run at the final cycle point
+        [[[ PT12H!$ ]]]
+            graph = baz
+
+        # Don't run at a speciffic datetime
```
Comment (hjoliver) [inline tests/cyclers/exclusions/suite.rc]: Too many 'f's! (specific)

**Comment by hjoliver** [comment]
We probably need to allow whitespace before the '!' character, or else detect it and abort with a sensible error message.  Currently I get, with `[[[P1Y !^]]]`:
```
cylc val  foo
Invalid ISO 8601 duration representation: P1Y
```

**Comment by hjoliver** [comment]
Would it be easy to extend this (via a new issue) to repeatedly exclude a given point in a sequence?   E.g. last week our ops team asked can they have a particular set of dependencies every 5 min but *not* on the hour (without writing out each 5 min point in separate graph heading).

**Comment by oliver-sanders** [comment]
@hjoliver 

- Whitespace now permitted `P1Y! ^`, `P1Y ! ^`, `P1Y !^`
- Fixed (embarrassing) time zone bug
- Improved docs
- Pruned reference logs

Excluding multiple points or indeed sequences is definitely possible and shouldn't prove too much work. I was thinking of creating an issue for something like:
- `[[[ PT3D ! PT9D ]]]`
- `[[[ T00 ! (20000101T00, 20100101T00, ...) ]]]`

File: doc/cug.tex [hunk 21]
```diff
@@ -2776,25 +2776,29 @@ \subsubsection{Graph Types}
 \label{fig-test2}
 \end{figure}
 
-\subsubsection{Graph Headers}
+\subsubsection{Graph Section Headings}
 
-In the above examples graph headings are used to define recurring date-times
-when tasks are to run. For example in the following scenario,
+Graph section headings define recurrence expressions, the graph within a graph
+section header defines a workflow at each point of the recurrence. For
```
Comment (hjoliver) [inline doc/cug.tex]: (Sorry to be pendantic, but!) can we have consistent use of "heading" rather than "header" throughout.

**Comment by hjoliver** [comment]
One more minor doc issue to address - "heading" instead of "header" (for some reason GitHub thinks my comment is already outdated?!).

The unit test is still failing here for what looks like a timezone problem (as I'm seeing 11-hour differences).  Maybe it needs `init(time_zone='Z')` or something?

```
$ python lib/cylc/cycling/iso8601.py
.F.
======================================================================
FAIL: test_exclusions_simple (__main__.TestISO8601Sequence)
Test the generation of points for sequences with exclusions.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "lib/cylc/cycling/iso8601.py", line 851, in test_exclusions_simple
    '20000101T0400', '20000101T0500'])
AssertionError: Lists differ: ['20000101T1200', '20000101T13... != ['20000101T0100', '20000101T02...

First differing element 0:
20000101T1200
20000101T0100

- ['20000101T1200', '20000101T1300', '20000101T1500', '20000101T1600']
?             -               ^^               ^^               ^^

+ ['20000101T0100', '20000101T0200', '20000101T0400', '20000101T0500']
?            +                ^^               ^^               ^^


----------------------------------------------------------------------
Ran 3 tests in 0.686s

FAILED (failures=1)

```

File: doc/cug.tex (hunk 21 above)
Comment (oliver-sanders) [inline doc/cug.tex]: Done.

**Comment by oliver-sanders** [comment]
@hjoliver Tests should pass now.

**Comment by hjoliver** [comment]
That's got it, thanks.

**Pull request merged**